cpp/src/community/detail/mis_impl.cuh

```diff
@@ -0,0 +1,729 @@
+
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <prims/fill_edge_src_dst_property.cuh>
+#include <prims/per_v_transform_reduce_incoming_outgoing_e.cuh>
+#include <prims/update_edge_src_dst_property.cuh>
+
+#include <community/detail/mis.hpp>
+#include <cugraph/edge_property.hpp>
+#include <cugraph/edge_src_dst_property.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/graph_view.hpp>
+
+#include <cugraph/utilities/host_scalar_comm.hpp>
+#include <cugraph/utilities/thrust_tuple_utils.hpp>
+
+#include <raft/util/cudart_utils.hpp>
+#include <raft/util/integer_utils.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cub/cub.cuh>
+#include <thrust/count.h>
+#include <thrust/distance.h>
+
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <thrust/fill.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/optional.h>
+#include <thrust/sequence.h>
+#include <thrust/shuffle.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+#include <thrust/transform_reduce.h>
+#include <thrust/tuple.h>
+#include <thrust/unique.h>
+
+#include <numeric>
+#include <type_traits>
+#include <utility>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+rmm::device_uvector<weight_t> rank_vertices(
+  raft::handle_t const& handle,
+  cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+  std::optional<cugraph::edge_property_view_t<edge_t, weight_t const*>> edge_weight_view)
+{
+  return compute_out_weight_sums<vertex_t, edge_t, weight_t, false, multi_gpu>(
+    handle, graph_view, *edge_weight_view);
+}
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+rmm::device_uvector<vertex_t> compute_mis(
+  raft::handle_t const& handle,
+  cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+  std::optional<cugraph::edge_property_view_t<edge_t, weight_t const*>> edge_weight_view)
+{
+  using GraphViewType = cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu>;
+  using FlagType      = vertex_t;
+
+  vertex_t number_of_vertices = graph_view.local_vertex_partition_range_size();
+
+  rmm::device_uvector<vertex_t> mis(number_of_vertices, handle.get_stream());
```

Comment:
`mis` is the return value and this variable is not used till very close to the end of this function. Defining this variable at the beginning wastes memory (you don't need to reserve this memory till the very end of this function) and also makes code more cryptic. A general rule is to minimize the scope of a variable. This makes code easier to understand and also avoid wasting memory (keep the peak memory usage low).